File: html5/src/ui/UiCanvas.js
// -----------------------------------------------------------------------------
// UiCanvas — one screen-space surface holding a tree of UiNode.
//
// The mirror of SexyBiscuit.Engine/UI/UiCanvas.cs, member for member. This is the
// retained tree that UiLayout, UiFocus and UiDocument work on, and what the `UI`
// script global builds into.
//
// The scale and offset computed here are used by *both* painting and hit-testing.
// The canvas this replaces had a scale matrix that nothing ever called, so pointer
// coordinates stayed in raw pixels and hit-testing quietly disagreed with what was
// on screen at every window size but one.
// -----------------------------------------------------------------------------

import { Component } from '../core/Component.js';
import { PropertyType as P } from '../core/PropertyTypes.js';
import { registerComponent } from '../core/TypeRegistry.js';
import { UiNode } from './UiNode.js';
import { UiKind, UiScaleMode, SafeAreaMode, UiSpace, UiFacing, PositionMode, canonical } from './UiEnums.js';
import { Vector3 } from '../math/Vector3.js';
import { Transform3D } from '../core/Transform3D.js';
import { Camera3D } from '../rendering/Camera3D.js';
import { build as buildBasis, rayToCanvas } from './UiWorld.js';
import { measure as measureLayout, arrange, hitTest as hitTestLayout, deflate } from './UiLayout.js';
import { UiInput } from './UiInput.js';
import { fromJson, fromObject, apply as uiApply, read as uiRead, WRITABLE_KEYS } from './UiDocument.js';

/** Every canvas that currently exists, for the host's screen-space pass. */
const all = [];

/**
 * Properties of the root that belong to the canvas, not to the document it adopts.
 *
 * A document's root is a box like any other and may well have been authored with a size and a
 * position; the canvas root is neither, it is the whole surface. Copying a width onto it would
 * shrink the UI to whatever the author happened to be looking at when they saved. Everything
 * else is copied -- by walking the codec's own key list rather than a hand-written one, so a
 * property added to the format is not quietly lost by a list nobody remembered to extend.
 */
const ROOT_OWNED_KEYS = new Set([
    'width', 'height', 'minwidth', 'minheight', 'maxwidth', 'maxheight',
    'grow', 'shrink', 'margin', 'positioning', 'anchor', 'anchormin', 'anchormax',
    'pivot', 'offset', 'offsetmax', 'worldfollow', 'worldanchor', 'worldfollowdistance',
]);

export class UiCanvas extends Component {
    /**
     * What a scene file stores: the canvas's settings and the path to its document,
     * never the tree. Mirrors the public properties of the C# UiCanvas, and
     * ComponentSchemaParityTests holds the two lists together.
     */
    static schema = {
        scaleMode:           { type: P.Enum, values: ['ConstantPixel', 'ScaleToFit', 'ScaleToFill', 'Match'], default: 'ConstantPixel' },
        referenceResolution: { type: P.Vector2, default: [1920, 1080] },
        matchWidthOrHeight:  { type: P.Number, default: 0.5, min: 0, max: 1 },
        order:               { type: P.Int, default: 0 },
        document:            { type: P.Asset, assetKind: 'ui', default: '' },
        safeArea:            { type: P.Vector4, default: [0, 0, 0, 0] },
        safeAreaMode:        { type: P.Enum, values: ['Ignore', 'Inset', 'InsetX', 'InsetY'], default: 'Ignore' },
        interactive:         { type: P.Bool, default: true },
        playerIndex:         { type: P.Int, default: -1 },
        space:               { type: P.Enum, values: ['Screen', 'World'], default: 'Screen' },
        facing:              { type: P.Enum, values: ['Billboard', 'VerticalBillboard', 'Plane'], default: 'Billboard' },
        worldPosition:       { type: P.Vector3, default: [0, 0, 0] },
        worldOffset:         { type: P.Vector3, default: [0, 0, 0] },
        pixelsPerUnit:       { type: P.Number, default: 100 },
        maxDrawDistance:     { type: P.Number, default: 0 },
        doubleSided:         { type: P.Bool, default: false },
        depthTest:           { type: P.Bool, default: true },
    };

    constructor() {
        super();
        // --- configuration ---
        this.scaleMode = UiScaleMode.ConstantPixel;
        this.referenceResolution = { x: 1920, y: 1080 };
        this.matchWidthOrHeight = 0.5;

        /** Paint order between canvases. Lower paints first, so higher is in front. */
        this.order = 0;

        /** Path to a `.ui` document. When set it replaces the tree at start. */
        this.document = '';

        /** Insets kept clear of notches and TV overscan, as left, top, right, bottom. */
        this.safeArea = { x: 0, y: 0, z: 0, w: 0 };
        this.safeAreaMode = SafeAreaMode.Ignore;

        /** Whether this canvas takes pointer and navigation input at all. */
        this.interactive = true;

        /** Which player drives this canvas, or -1 for anybody's input. */
        this.playerIndex = -1;

        // --- world space ---

        /** Whether this canvas is a surface on the screen or a plane in the world. */
        this.space = UiSpace.Screen;

        /** How a world-space canvas is turned to face the player. */
        this.facing = UiFacing.Billboard;

        /** Where a world canvas stands when its actor has no Transform3D -- or no actor. */
        this.worldPosition = { x: 0, y: 0, z: 0 };

        /** Shifts a world canvas off its anchor, in world units. */
        this.worldOffset = { x: 0, y: 0, z: 0 };

        /** Canvas units per world unit: the size dial for a world canvas. */
        this.pixelsPerUnit = 100;

        /** Stop drawing a world canvas past this distance. Zero never stops. */
        this.maxDrawDistance = 0;

        /** Whether a world canvas is legible from behind as well as in front. */
        this.doubleSided = false;

        /** Whether geometry in front of a world canvas hides it. */
        this.depthTest = true;

        /** The last canvas point a pointer ray found, held across a drag that leaves the plane. */
        this._lastWorldPoint = null;

        /**
         * Whether any node in this tree has ever asked to follow a world point.
         *
         * Latched rather than counted. A count would have to be kept correct across every
         * add, remove, reparent and property write, and getting that wrong shows up as a
         * marker that silently stops moving; a latch can only cost a walk that finds nothing.
         */
        this._hasWorldFollowers = false;

        /** The tree. Always present; an empty root paints nothing. */
        this.root = new UiNode({ name: 'Root', kind: UiKind.Panel });
        this.root.canvas = this;

        // --- resolved each frame ---
        this.scale = { x: 1, y: 1 };
        this.canvasOffset = { x: 0, y: 0 };
        this.canvasSize = { x: 1280, y: 720 };

        this._viewport = { x: 1280, y: 720 };
        this._input = null;

        all.push(this);
    }

    /**
     * A point far enough outside any canvas that every rectangle test rejects it.
     *
     * Finite on purpose. An infinity or a NaN would propagate through the slider arithmetic
     * in UiInput instead of simply missing, and "missing" is exactly what a pointer ray that
     * does not meet a canvas's plane should mean.
     */
    static NOWHERE = Object.freeze({ x: -1000000, y: -1000000 });

    /** Every live canvas, in creation order. */
    static get all() { return all; }

    /** Drops every canvas. A test that leaks one poisons the next. */
    static clearAll() { all.length = 0; }

    /** Removes this canvas from the paint list. */
    /**
     * Deregisters the canvas, so the host stops laying it out and painting it.
     *
     * Named for the Component hook the scene calls, which is what the C# side
     * overrides too; `destroy()` stays as the spelling direct owners already use.
     */
    onDestroy() {
        const i = all.indexOf(this);
        if (i >= 0) all.splice(i, 1);
    }

    destroy() { this.onDestroy(); }

    // -------------------------------------------------------------------------
    // Input
    // -------------------------------------------------------------------------

    /**
     * This canvas's pointer, focus and control behaviour. Created on first use,
     * because the router holds focus state that has to survive between frames.
     */
    get input() {
        if (!this._input) this._input = new UiInput(this);
        return this._input;
    }

    // -------------------------------------------------------------------------
    // Scale
    // -------------------------------------------------------------------------

    /** Recomputes the scale and offset for a viewport size. Call before laying out. */
    setViewport(width, height) {
        const w = Math.max(1, width);
        const h = Math.max(1, height);
        if (this._viewport.x === w && this._viewport.y === h) return;

        this._viewport = { x: w, y: h };
        this.root.invalidateMeasure();
    }

    _resolveScale() {
        // A world canvas is not fitted to anything: it is a fixed sheet standing in the scene,
        // so its size is what it was authored at and its transform is the identity. That one
        // branch is why every other thing in the UI -- layout, painting, clipping, focus,
        // navigation, hit testing -- keeps working in world space without being told about it.
        if (this.space === UiSpace.World) {
            this.scale = { x: 1, y: 1 };
            this.canvasOffset = { x: 0, y: 0 };
            this.canvasSize = {
                x: Math.max(1, this.referenceResolution.x),
                y: Math.max(1, this.referenceResolution.y),
            };
            return;
        }

        const vw = this._viewport.x;
        const vh = this._viewport.y;
        const rw = Math.max(1, this.referenceResolution.x);
        const rh = Math.max(1, this.referenceResolution.y);

        switch (this.scaleMode) {
            case UiScaleMode.ScaleToFit: {
                const s = Math.min(vw / rw, vh / rh);
                this.scale = { x: s, y: s };
                this.canvasSize = { x: rw, y: rh };
                break;
            }
            case UiScaleMode.ScaleToFill: {
                const s = Math.max(vw / rw, vh / rh);
                this.scale = { x: s, y: s };
                this.canvasSize = { x: rw, y: rh };
                break;
            }
            case UiScaleMode.Match: {
                const m = Math.min(1, Math.max(0, this.matchWidthOrHeight));
                const s = Math.pow(vw / rw, 1 - m) * Math.pow(vh / rh, m);
                this.scale = { x: s, y: s };
                this.canvasSize = { x: vw / s, y: vh / s };
                break;
            }
            default:
                this.scale = { x: 1, y: 1 };
                this.canvasSize = { x: vw, y: vh };
                break;
        }

        // Centre whatever the scaled canvas does not cover, so a letterbox is even.
        this.canvasOffset = {
            x: (vw - this.canvasSize.x * this.scale.x) * 0.5,
            y: (vh - this.canvasSize.y * this.scale.y) * 0.5,
        };
    }

    // -------------------------------------------------------------------------
    // Coordinates
    // -------------------------------------------------------------------------

    /**
     * Device pixels to canvas units. Every hit test starts here.
     *
     * The single seam between the two spaces, which is why world canvases cost the input
     * router nothing: it still asks one question and still gets canvas units back.
     */
    screenToCanvas(point) {
        if (this.space === UiSpace.World) return this._worldScreenToCanvas(point);

        return {
            x: (point.x - this.canvasOffset.x) / this.scale.x,
            y: (point.y - this.canvasOffset.y) / this.scale.y,
        };
    }

    _worldScreenToCanvas(point) {
        const basis = this.worldBasis();
        const camera = Camera3D.main;

        if (basis && camera) {
            const ray = camera.screenToWorldRay(point, this._viewport.x, this._viewport.y);
            const hit = rayToCanvas(basis, ray.origin, ray.direction, this.canvasSize);
            if (hit) {
                this._lastWorldPoint = hit;
                return hit;
            }
        }

        // A drag holds its last good point rather than reporting a miss. Swinging the camera
        // until the ray leaves the plane would otherwise snap the slider you are dragging to
        // its minimum, which is a worse answer than "wherever you last had it".
        if (this._input?.isDragging && this._lastWorldPoint) return this._lastWorldPoint;

        return { ...UiCanvas.NOWHERE };
    }

    // -------------------------------------------------------------------------
    // World space
    // -------------------------------------------------------------------------

    /** The actor's 3D transform when it has one; a script's canvas has no actor at all. */
    get transform3D() { return this.actor?.getComponent(Transform3D) ?? null; }

    /** The centre of a world canvas: its actor's position plus the offset. */
    get worldAnchor() {
        const t = this.transform3D;
        const origin = t ? t.position : this.worldPosition;
        return Vector3.add(Vector3.from(origin), Vector3.from(this.worldOffset));
    }

    /** The canvas's extent in world units: its size in canvas units, scaled down. */
    get worldSize() {
        const ppu = Math.max(0.0001, this.pixelsPerUnit);
        return { x: this.canvasSize.x / ppu, y: this.canvasSize.y / ppu };
    }

    /** The plane this canvas occupies, or null when it is not in the world or has no camera. */
    worldBasis() {
        if (this.space !== UiSpace.World) return null;

        const camera = Camera3D.main;
        if (!camera) return null;

        const t = this.transform3D;

        return buildBasis(
            this.worldAnchor,
            this.facing,
            t ? t.rotation : { x: 0, y: 0, z: 0, w: 1 },
            camera.getViewMatrix(),
            camera.getTransform3D().position,
            this.worldSize);
    }

    /** Whether a world canvas is close enough to the camera to be worth drawing. */
    get isWithinDrawDistance() {
        if (this.maxDrawDistance <= 0) return true;

        const camera = Camera3D.main;
        if (!camera) return true;

        const away = Vector3.subtract(camera.getTransform3D().position, this.worldAnchor);
        return away.lengthSquared <= this.maxDrawDistance * this.maxDrawDistance;
    }

    /** Canvas units to device pixels. Clip rectangles must go through this. */
    canvasToScreen(point) {
        return {
            x: point.x * this.scale.x + this.canvasOffset.x,
            y: point.y * this.scale.y + this.canvasOffset.y,
        };
    }

    /** A canvas rectangle in device pixels, for the clip region. */
    canvasRectToScreen(r) {
        const topLeft = this.canvasToScreen({ x: r.x, y: r.y });
        const bottomRight = this.canvasToScreen({ x: r.x + r.width, y: r.y + r.height });
        return {
            x: topLeft.x,
            y: topLeft.y,
            width: bottomRight.x - topLeft.x,
            height: bottomRight.y - topLeft.y,
        };
    }

    // -------------------------------------------------------------------------
    // Layout
    // -------------------------------------------------------------------------

    /**
     * The rectangle the root is laid out into: the canvas, less the safe-area
     * insets on whichever axes they apply to.
     */
    get safeRect() {
        const full = { x: 0, y: 0, width: this.canvasSize.x, height: this.canvasSize.y };
        const s = this.safeArea;

        switch (this.safeAreaMode) {
            case SafeAreaMode.Inset:  return deflate(full, s);
            case SafeAreaMode.InsetX: return deflate(full, { x: s.x, y: 0, z: s.z, w: 0 });
            case SafeAreaMode.InsetY: return deflate(full, { x: 0, y: s.y, z: 0, w: s.w });
            default:                  return full;
        }
    }

    /**
     * Measures and arranges the tree if anything has changed since the last pass.
     * Cheap to call every frame; that is the point of the dirty flags.
     */
    layout() {
        this._resolveScale();
        this._resolveWorldFollowers();
        if (!this.root.measureDirty && !this.root.arrangeDirty) return;

        const area = this.safeRect;
        measureLayout(this.root, { x: area.width, y: area.height });
        arrange(this.root, area, { x: 0, y: 0, width: this.canvasSize.x, height: this.canvasSize.y });
    }

    /** Tells this canvas one of its nodes wants to follow a world point. */
    noteWorldFollower() { this._hasWorldFollowers = true; }

    /**
     * Moves every following node to wherever its world anchor is on screen this frame.
     *
     * Runs before the dirty check rather than after it, because a node that follows a moving
     * actor is dirty by definition and would otherwise be laid out once and left. A canvas
     * with no followers pays one boolean.
     *
     * Nodes behind the camera are hidden rather than placed: the perspective divide flips a
     * point behind the viewer to the opposite side of the screen, which is why worldToScreen
     * returns null there instead of a coordinate.
     */
    _resolveWorldFollowers() {
        if (!this._hasWorldFollowers) return;

        const camera = Camera3D.main;
        const eye = camera ? camera.getTransform3D().position : null;

        for (const node of this.root.descendants()) {
            if (!node.worldFollow) continue;

            node.positioning = PositionMode.Absolute;

            if (!camera) {
                node.visible = false;
                continue;
            }

            if (node.worldFollowDistance > 0) {
                const away = Vector3.subtract(Vector3.from(eye), Vector3.from(node.worldAnchor));
                if (away.lengthSquared > node.worldFollowDistance * node.worldFollowDistance) {
                    node.visible = false;
                    continue;
                }
            }

            const screen = camera.worldToScreen(node.worldAnchor, this._viewport.x, this._viewport.y);
            if (!screen) {
                node.visible = false;
                continue;
            }

            node.visible = true;
            node.offset = this.screenToCanvas(screen);
        }
    }

    /** Forces a full pass next frame, whatever the dirty flags say. */
    invalidateLayout() { this.root.invalidateMeasure(); }

    // -------------------------------------------------------------------------
    // Hit testing
    // -------------------------------------------------------------------------

    /** The top-most interactive node under a device-pixel point, or null. */
    hitTest(screenPoint) {
        if (!this.interactive) return null;
        return hitTestLayout(this.root, this.screenToCanvas(screenPoint));
    }

    /** Finds a node by name anywhere in this canvas. */
    find(name) { return this.root.find(name); }

    // -------------------------------------------------------------------------
    // Documents
    // -------------------------------------------------------------------------

    /**
     * Loads the `.ui` document named by `document`, if there is one.
     *
     * Asynchronous, and the one real asymmetry with the C# side, which reads the file inside
     * a synchronous Awake. A browser cannot; so the canvas paints empty for the frame or two
     * the fetch takes and then fills in. Before this existed the browser never loaded the
     * property at all -- it round-tripped through a `.scene` file and did nothing -- so a UI
     * saved in the editor appeared natively and nowhere else.
     */
    awake() {
        super.awake?.();
        if (!this.document) return;

        // The same route SpriteRenderer takes to its texture, rather than a second one.
        const assets = this.actor?.scene?.engine?.assets;
        if (!assets) return;

        assets.loadJson(this.document)
            .then((spec) => { if (spec) this.adopt(fromObject(spec)); })
            .catch((error) => {
                console.warn(`[UiCanvas] could not load document '${this.document}': ${error.message}`);
            });
    }

    /** Parses a `.ui` document and adopts it. */
    loadDocument(source) { this.adopt(fromJson(source)); }

    /**
     * Moves a loaded document's children onto the root, replacing what was there.
     *
     * The document's own root is not adopted as this canvas's root — its layout
     * properties are copied across and its children are moved. A canvas whose root
     * could be swapped would break every reference the host and the router hold.
     */
    adopt(document) {
        for (let i = this.root.children.length - 1; i >= 0; i--) {
            this.root.remove(this.root.children[i]);
        }

        for (const key of WRITABLE_KEYS) {
            if (ROOT_OWNED_KEYS.has(canonical(key))) continue;
            uiApply(this.root, key, uiRead(document, key));
        }

        for (const child of [...document.children]) this.root.add(child);
        this.invalidateLayout();
    }
}

registerComponent(UiCanvas, {
    category: 'UI',
    fullName: 'SexyBiscuit.Engine.UI.UiCanvas',
    summary: 'A screen-space surface holding a tree of UI nodes.',
});

File: SexyBiscuit.Engine/UI/UiCanvas.cs
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using SexyBiscuit.Engine.Core;

namespace SexyBiscuit.Engine.UI;

/// <summary>
/// One screen-space surface holding a tree of <see cref="UiNode"/>. Attach it to an
/// Actor to put a UI in a scene.
/// </summary>
/// <remarks>
/// <para>
/// A canvas does not paint through <see cref="Component.Draw"/>, and that is deliberate.
/// Component drawing happens inside the camera-transformed world batch, which is why the
/// UI this replaces panned, zoomed and shook with the camera and landed at the back of a
/// front-to-back sort. The host collects canvases and paints them in the screen-space pass
/// that runs after the world, in <see cref="Order"/>.
/// </para>
/// <para>
/// The scale and offset computed here are used by <em>both</em> painting and hit-testing.
/// The canvas this replaces had a public, documented scale matrix that nothing ever called,
/// so pointer coordinates stayed in raw screen pixels and hit-testing quietly disagreed with
/// what was on screen at any window size but one.
/// </para>
/// <para>
/// The canvas may have no Actor: the process-wide canvas behind the script <c>UI</c> global
/// is constructed directly. Nothing here may dereference <see cref="Component.Actor"/>.
/// </para>
/// </remarks>
public class UiCanvas : Component
{
    // -------------------------------------------------------------------------
    // Every live canvas, in paint order
    // -------------------------------------------------------------------------

    private static readonly List<UiCanvas> _all = new();

    /// <summary>Every canvas that currently exists, for the host's screen-space pass.</summary>
    public static IReadOnlyList<UiCanvas> All => _all;

    public UiCanvas()
    {
        _all.Add(this);
        Root.Canvas = this;
    }

    public override void OnDestroy()
    {
        _all.Remove(this);
        base.OnDestroy();
    }

    /// <summary>Drops every canvas. A test that leaks one poisons the next.</summary>
    internal static void ClearAll() => _all.Clear();

    // -------------------------------------------------------------------------
    // Configuration
    // -------------------------------------------------------------------------

    /// <summary>How the reference resolution maps onto the viewport.</summary>
    public UiScaleMode ScaleMode { get; set; } = UiScaleMode.ConstantPixel;

    /// <summary>The resolution the tree is authored at, for every mode but ConstantPixel.</summary>
    public Vector2 ReferenceResolution { get; set; } = new(1920f, 1080f);

    /// <summary>Blends the width and height ratios under <see cref="UiScaleMode.Match"/>.</summary>
    public float MatchWidthOrHeight { get; set; } = 0.5f;

    /// <summary>Paint order between canvases. Lower paints first, so higher is in front.</summary>
    public int Order { get; set; }

    /// <summary>Path to a <c>.ui</c> document. When set it replaces <see cref="Root"/> at start.</summary>
    public string Document { get; set; } = "";

    /// <summary>Insets kept clear of notches and TV overscan, as left, top, right, bottom.</summary>
    public Vector4 SafeArea { get; set; }

    /// <summary>Which axes <see cref="SafeArea"/> applies to.</summary>
    public SafeAreaMode SafeAreaMode { get; set; } = SafeAreaMode.Ignore;

    /// <summary>Whether this canvas takes pointer and navigation input at all.</summary>
    public bool Interactive { get; set; } = true;

    /// <summary>Which player drives this canvas, or -1 for anybody's input.</summary>
    public int PlayerIndex { get; set; } = -1;

    /// <summary>The tree. Always present; an empty root paints nothing.</summary>
    public UiNode Root { get; } = new() { Name = "Root", Kind = UiKind.Panel };

    // -------------------------------------------------------------------------
    // Resolved each frame
    // -------------------------------------------------------------------------

    /// <summary>Canvas units to device pixels.</summary>
    public Vector2 Scale { get; private set; } = Vector2.One;

    /// <summary>Device-pixel offset of the canvas origin, which centres a letterbox.</summary>
    public Vector2 CanvasOffset { get; private set; }

    /// <summary>The viewport in canvas units, which is what a full-screen node fills.</summary>
    public Vector2 CanvasSize { get; private set; } = new(1280f, 720f);

    private Vector2 _viewport = new(1280f, 720f);

    /// <summary>
    /// Recomputes the scale and offset for a viewport size. Call before laying out.
    /// </summary>
    public void SetViewport(float width, float height)
    {
        width  = MathF.Max(1f, width);
        height = MathF.Max(1f, height);

        if (_viewport.X == width && _viewport.Y == height) return;
        _viewport = new Vector2(width, height);
        Root.InvalidateMeasure();
    }

    private void ResolveScale()
    {
        float vw = _viewport.X, vh = _viewport.Y;
        float rw = MathF.Max(1f, ReferenceResolution.X);
        float rh = MathF.Max(1f, ReferenceResolution.Y);

        switch (ScaleMode)
        {
            case UiScaleMode.ScaleToFit:
            {
                float s = MathF.Min(vw / rw, vh / rh);
                Scale = new Vector2(s, s);
                CanvasSize = new Vector2(rw, rh);
                break;
            }
            case UiScaleMode.ScaleToFill:
            {
                float s = MathF.Max(vw / rw, vh / rh);
                Scale = new Vector2(s, s);
                CanvasSize = new Vector2(rw, rh);
                break;
            }
            case UiScaleMode.Match:
            {
                float m = Math.Clamp(MatchWidthOrHeight, 0f, 1f);
                float s = MathF.Pow(vw / rw, 1f - m) * MathF.Pow(vh / rh, m);
                Scale = new Vector2(s, s);
                CanvasSize = new Vector2(vw / s, vh / s);
                break;
            }
            default:
                Scale = Vector2.One;
                CanvasSize = new Vector2(vw, vh);
                break;
        }

        // Centre whatever the scaled canvas does not cover, so a letterbox is even.
        CanvasOffset = new Vector2(
            (vw - CanvasSize.X * Scale.X) * 0.5f,
            (vh - CanvasSize.Y * Scale.Y) * 0.5f);
    }

    // -------------------------------------------------------------------------
    // Coordinates
    // -------------------------------------------------------------------------

    /// <summary>Device pixels to canvas units. Every hit test starts here.</summary>
    public Vector2 ScreenToCanvas(Vector2 screen) => new(
        (screen.X - CanvasOffset.X) / Scale.X,
        (screen.Y - CanvasOffset.Y) / Scale.Y);

    /// <summary>Canvas units to device pixels. Scissor rectangles must go through this.</summary>
    public Vector2 CanvasToScreen(Vector2 canvas) => new(
        canvas.X * Scale.X + CanvasOffset.X,
        canvas.Y * Scale.Y + CanvasOffset.Y);

    /// <summary>
    /// A canvas rectangle in device pixels, for the scissor test.
    /// </summary>
    /// <remarks>
    /// <c>GraphicsDevice.ScissorRectangle</c> is in device pixels and is not affected by
    /// the batch transform, so a clip rectangle passed straight through would be wrong by
    /// exactly the canvas scale.
    /// </remarks>
    public RectangleF CanvasToScreen(RectangleF r)
    {
        Vector2 topLeft     = CanvasToScreen(new Vector2(r.X, r.Y));
        Vector2 bottomRight = CanvasToScreen(new Vector2(r.Right, r.Bottom));
        return new RectangleF(topLeft.X, topLeft.Y, bottomRight.X - topLeft.X, bottomRight.Y - topLeft.Y);
    }

    /// <summary>The transform to hand <c>SpriteBatch.Begin</c> so canvas units paint correctly.</summary>
    public Matrix GetTransform()
        => Matrix.CreateScale(Scale.X, Scale.Y, 1f)
         * Matrix.CreateTranslation(CanvasOffset.X, CanvasOffset.Y, 0f);

    // -------------------------------------------------------------------------
    // Layout
    // -------------------------------------------------------------------------

    /// <summary>
    /// The rectangle the root is laid out into: the canvas, less the safe-area insets
    /// on whichever axes they apply to.
    /// </summary>
    public RectangleF SafeRect
    {
        get
        {
            var full = new RectangleF(0f, 0f, CanvasSize.X, CanvasSize.Y);
            return SafeAreaMode switch
            {
                SafeAreaMode.Inset  => full.Deflate(SafeArea),
                SafeAreaMode.InsetX => full.Deflate(new Vector4(SafeArea.X, 0f, SafeArea.Z, 0f)),
                SafeAreaMode.InsetY => full.Deflate(new Vector4(0f, SafeArea.Y, 0f, SafeArea.W)),
                _                   => full,
            };
        }
    }

    /// <summary>
    /// Measures and arranges the tree if anything has changed since the last pass.
    /// Cheap to call every frame; that is the point of the dirty flags.
    /// </summary>
    public void Layout()
    {
        ResolveScale();
        if (!Root.MeasureDirty && !Root.ArrangeDirty) return;

        RectangleF area = SafeRect;
        UiLayout.Measure(Root, new Vector2(area.Width, area.Height));
        UiLayout.Arrange(Root, area, new RectangleF(0f, 0f, CanvasSize.X, CanvasSize.Y));
    }

    /// <summary>Forces a full pass next frame, whatever the dirty flags say.</summary>
    public void InvalidateLayout() => Root.InvalidateMeasure();

    // -------------------------------------------------------------------------
    // Hit testing
    // -------------------------------------------------------------------------

    /// <summary>
    /// The top-most interactive node under a device-pixel point, or null.
    /// </summary>
    public UiNode? HitTest(Vector2 screenPoint)
    {
        if (!Interactive) return null;
        return UiLayout.HitTest(Root, ScreenToCanvas(screenPoint));
    }

    /// <summary>Finds a node by name anywhere in this canvas.</summary>
    public UiNode? Find(string name) => Root.Find(name);

    // -------------------------------------------------------------------------
    // Input
    // -------------------------------------------------------------------------

    private UiInput? _input;

    /// <summary>
    /// This canvas's pointer, focus and control behaviour. Created on first use.
    /// </summary>
    /// <remarks>
    /// Lazy because a canvas that is only ever painted — a pure HUD — should not pay for
    /// a focus model it never consults, and because the router holds the focus state that
    /// has to survive between frames, so it cannot be created per frame by the host.
    /// </remarks>
    public UiInput Input => _input ??= new UiInput(this);

    // -------------------------------------------------------------------------
    // Documents
    // -------------------------------------------------------------------------

    /// <summary>Loads <see cref="Document"/>, if one is named, into <see cref="Root"/>.</summary>
    public override void Awake()
    {
        base.Awake();
        LoadDocument();
    }

    /// <summary>
    /// Replaces the tree with the contents of a <c>.ui</c> file.
    /// </summary>
    /// <remarks>
    /// The document's own root is not adopted as this canvas's root — its layout properties
    /// are copied onto <see cref="Root"/> and its children are moved across. A canvas whose
    /// root could be swapped would break every reference the host and the router hold, and
    /// <see cref="Root"/> is deliberately get-only for that reason.
    /// </remarks>
    public void LoadDocument()
    {
        if (string.IsNullOrWhiteSpace(Document)) return;

        string path = Core.ProjectPaths.Resolve(Document);
        if (!File.Exists(path)) return;

        Adopt(UiDocument.FromJson(File.ReadAllText(path)));
    }

    /// <summary>Moves a loaded document's children onto <see cref="Root"/>, replacing what was there.</summary>
    public void Adopt(UiNode document)
    {
        ArgumentNullException.ThrowIfNull(document);

        for (int i = Root.Children.Count - 1; i >= 0; i--) Root.Remove(Root.Children[i]);

        Root.Layout       = document.Layout;
        Root.Gap          = document.Gap;
        Root.Wrap         = document.Wrap;
        Root.Padding      = document.Padding;
        Root.MainAlign    = document.MainAlign;
        Root.CrossAlign   = document.CrossAlign;
        Root.Columns      = document.Columns;
        Root.CellSize     = document.CellSize;
        Root.Background   = document.Background;
        Root.Clip         = document.Clip;
        Root.Scroll       = document.Scroll;
        Root.ScrollOffset = document.ScrollOffset;

        foreach (UiNode child in new List<UiNode>(document.Children)) Root.Add(child);
        InvalidateLayout();
    }
}
